runtime/commonMain/src/kotlinx/serialization/SerialDescriptorBuilder.kt

```diff
@@ -225,16 +241,12 @@ private class SerialDescriptorImpl(
     override fun getElementDescriptor(index: Int): SerialDescriptor = elementDescriptors.getChecked(index)
     override fun isElementOptional(index: Int): Boolean = elementOptionality.getChecked(index)
 
-    override fun equals(other: Any?): Boolean {
-        if (this === other) return true
-        if (other !is SerialDescriptor) return false
-        if (serialName != other.serialName) return false
-        return true
-    }
+    override fun equals(other: Any?): Boolean =
+        equalsImpl(other) { otherDescriptor: SerialDescriptorImpl -> typeParameters == otherDescriptor.typeParameters }
 
-    override fun hashCode(): Int {
-        return serialName.hashCode()
-    }
+    private val _hashCode: Int by lazy { hashCodeImpl(typeParameters) }
```

Comment:
1) Please always declare `val`s in the beginning of the class, not in the middle of it.
2) For the idempotent computations, please prefer non-synchronized approach with zero-check similarly to `java.lang.String.hashCode`. It has a smaller footprint and better performance characteristics

Reply:
Non-synchronized approach is a good idea; however, I have a feeling that such an approach may fail in Native, if such a descriptor would be transferred between workers and frozen. Note that standard `lazy` is freeze-aware and serializers which are `object`s (w/o type params) are already frozen by default (https://github.com/JetBrains/kotlin-native/issues/1817).

It is also possible to simply not to cache hashcode if `this.isFrozen == true`, but I'm not sure that this is a great idea.

Reply:
Oh, okay. Let'sdrop a comment then, maybe at some point, we'll introduce some platform-specific class for that